### PR TITLE
Add Unit Test to Thoroughly Tests Mastodon's Undo Redo Mechanism

### PR DESCRIPTION
**Pull request by maarzt**
There is a PR to fix a small problem in Mastodon's undo/redo mechanism: https://github.com/mastodon-sc/mastodon-collection/pull/16

The goal of this PR is to ensure that Mastodons very important undo/redo mechanism works properly. I there for added a unit test `ModelUndoRedoTest`. The unit test uses the `ModelUtils.dump(...)` method to convert the current status of the `Model` into a string. Converting the model status into string allows me to test if the model is correctly rolled back into a specific (previously) recorded state.

It was necessary to extend the `ModelUtils.dump(...)` method to also print the tag sets. (The undo redo mechanism handle tags as well.)

### CODE REVIEW

**Comment by maarzt** [comment]
The test currently fails because of a but in the Undo/Redo mechanism. That will be fixed by this PR https://github.com/mastodon-sc/mastodon-collection/pull/16

File: src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+	/**
+	 * This method makes various changes to the graph and the tag set model. The
+	 * changes are divided into several steps, and at each step, an undo point is
+	 * set. Then, the undo and redo methods are called to test if the model is
+	 * restored to the intended state.
+	 */
+	private static void testUndoRedo( Model model )
+	{
+		final List< String > states = new ArrayList<>();
+		makeVariousChangesAndRecordUndoPoints( model, states );
```
Comment (stefanhahmann) [inline src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java:125]: I find it confusing when reading code, where changes to the variables that are given to the method happen inside the method. I usually do not expect this to happen. 
Could this method maybe return a Pair<Model, List> instead? That would confuse me less.

File: src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+		testUndoRedo( model );
+	}
+
+	/**
+	 * This method makes various changes to the graph and the tag set model. The
+	 * changes are divided into several steps, and at each step, an undo point is
+	 * set. Then, the undo and redo methods are called to test if the model is
+	 * restored to the intended state.
+	 */
+	private static void testUndoRedo( Model model )
```
Comment (stefanhahmann) [inline src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java:122]:
```suggestion
	private static void testUndoRedo( final Model model )
```
It seems like you forgot the final modifiers for the variables. It is not strictly speaking required, but seems to be a convention in this project.

File: src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	 */
+	@Test
+	public void testUndoRedoAfterModelImporter()
+	{
+		Model model = new Model();
+		// Fill the undo-redo history with many different entries:
+		makeVariousChangesAndRecordUndoPoints( model, new ArrayList<>() );
+		addFourSpotGraph( model );
+		// Reset the model by running the ModelImporter / Simulate an import operation:
+		new ModelImporter( model )
```
Comment (stefanhahmann) [inline src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java:230]: This is difficult to read and understand what is happening and why it is written in this way.

However, the constructor of ModelImporter is protected and should not be made public only for this test thus I do not see a simpler alternative to this.

**Review by stefanhahmann — CHANGES_REQUESTED** [review]
The main contribution of this PR is a unit test that checks if the undo/redo function works correctly on the Model. This is very valuable! Some parts of the test need some time to understand them, but overall the test is well written.
Also, this PR improves the dumping methods of ModelUtils a bit, which is also a good contribution.

I made some suggestion mainly regarding code formatting and also a bit regarding code structuring.

File: src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java (hunk 1 above)
Comment (maarzt) [reply src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java]: I introduced a UndoVerifier class to solve that problem.

File: src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java (hunk 3 above)
Comment (maarzt) [reply src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java]: There is now a Importer subclass that avoids this problem.

File: src/test/java/org/mastodon/mamut/model/ModelUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	{
+		Model model = new Model();
+		ModelGraph graph = model.getGraph();
+		Spot a = graph.addVertex().init( 0, new double[] { 1, 2, 3 }, 1 );
+		Spot b = graph.addVertex().init( 1, new double[] { 1, 2, 3.2 }, 1 );
+		a.setLabel( "A" );
+		b.setLabel( "B" );
+		graph.addEdge( a, b ).init();
+		String actual = ModelUtils.dump( model );
+		String expexted = "Model " + model + "\n"
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/test/java/org/mastodon/mamut/model/ModelUtilsTest.java]: done

File: src/test/java/org/mastodon/mamut/model/ModelUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+				+ "-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n"
+				+ "        0          A       0        1.0        2.0        3.0                   0                   1               1             0        1.0        2.0        3.0            1.0\n"
+				+ "        1          B       1        1.0        2.0        3.2                   1                   0               1             1        1.0        2.0        3.2            1.0\n"
+				+ "Links:\n"
+				+ "       Id  Source Id  Target Id    Link delta T    Link displacement    Source spot id    Target spot id    Link velocity\n"
+				+ "                                                             (pixel)                                        (pixel/frame)\n"
+				+ "-------------------------------------------------------------------------------------------------------------------------\n"
+				+ "        0          0          1             1.0                  0.2               0.0               1.0              0.2\n";
+		assertEquals( expexted, actual );
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/test/java/org/mastodon/mamut/model/ModelUtilsTest.java:69]: done

**Review by maarzt — COMMENTED** [review]
I addressed all the reviewers comments

File: src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java (hunk 1 above)
Comment (stefanhahmann) [reply src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java]: This is better now. Perhaps `UndoRedoModifier` would be a bit better suited as class name?

File: src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java (hunk 3 above)
Comment (stefanhahmann) [reply src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java]: Good idea. I like it.

File: src/main/java/org/mastodon/mamut/model/ModelUtils.java [hunk 6]
```diff
@@ -65,231 +74,233 @@ public static final String dump( final Model model )
...
+		if ( projection.isSet( t ) )
+			if ( projection instanceof IntFeatureProjection )
+				return String.format( "%" + width + "d", ( int ) projection.value( t ) );
+			else
+				return String.format( "%" + width + ".1f", projection.value( t ) );
+		else
+			return String.format( "%" + width + "s", "unset" );
+	}
+
+	private static class TablePrinter< T >
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/model/ModelUtils.java:252]: Okay.

File: src/main/java/org/mastodon/mamut/model/ModelUtils.java [hunk 7]
```diff
@@ -65,231 +74,233 @@ public static final String dump( final Model model )
...
-	 * 
-	 * @param model
-	 *            the model.
-	 * @param maxLines
-	 *            the max number of rows to print in the two tables.
-	 * @return a String representation of the model.
+	 * Which information to include in the table can be changed using
+	 * {@link DumpFlags}.
+	 */
+	public static final String dump( final Model model, final DumpFlags... options )
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/model/ModelUtils.java:99]: I see. Thanks for the explanation. The link is an interesting read. There are pros and cons regarding this that I was not aware of.

**Review by stefanhahmann — APPROVED** [review]
Apart from a minor suggestion to rename UndoVerifier to UndoRedoVerifier, which I leave up to your decision, I have no further suggestions to this PR.

**Pull request merged**